File: Python Impl/Code.py
import sys
import collections

"""
The Code class has all the methods for simulating the cd command of UNIX.
"""


class Commandline(object):
    """
    The constructor instantiates an object of the code class with the current directory as one of
    the parameters.
    """

    def __init__(self, directory):
        if directory is None or directory == "":
            raise Exception("invalid current directory")
        self.current_directory = directory
        self.new_directory = ""
        self.map = self.load_map()

    def load_map(self):
        map = collections.defaultdict()
        map[""] = self.process_empty
        map["."] = self.process_period
        map[".."] = self.process_parent
        return map

    def process(self, c_string):
        if c_string == "":
            self.new_directory = self.current_directory
            return self.new_directory
        c_string = self._set_new_directory(c_string)
        lst = c_string.split('/') if '/' in c_string else ["", c_string]
        for i in range(1, len(lst)):
            function = self.map.get(lst[i])
            if function:
                function()
            else:
                if not lst[i].isalnum():
                    raise Exception("No such file or Directory")
                self.process_directory(lst[i])
        return self.new_directory

    def _set_new_directory(self, c_string):
        if c_string.startswith('../') or c_string == '..':
            directory = '/' + self.current_directory.split('/')[-1]
            self.new_directory = self.current_directory[:-len(directory)]
            if '/' not in self.new_directory:
                self.new_directory = '/'
            c_string = c_string.replace("..", "", 1)
        elif c_string.startswith('./') or c_string == '.':
            self.new_directory = self.current_directory
            c_string = c_string.replace(".", "", 1)
        elif c_string.startswith('/'):
            self.new_directory = '/'
        else:
            self.new_directory = self.current_directory
        return c_string

    def process_empty(self):
        pass

    def process_period(self):
        pass

    def process_parent(self):
        directory = '/' + self.new_directory.split('/')[-1]
        self.new_directory = self.new_directory[:-len(directory)]
        if '/' not in self.new_directory:
            self.new_directory = '/'

    def process_directory(self, directory):
        if self.new_directory[-1] != '/':
            self.new_directory += '/'
        self.new_directory += directory


if __name__ == "__main__":
    current_directory = ""
    command_string = ""
    current_directory = sys.argv[1]
    command_string = sys.argv[2]
    command = Commandline(current_directory)
    print(command.process(command_string))

File: Java_Impl/src/Code.java
import java.util.HashMap;
import java.util.Map;

/**
 * The Code class has all the methods for simulating the cd command of UNIX.
 */
public class Code {

  private final String currentDirectory;
  private String newDirectory;
  private final Map<String, Runnable> map;

  /**
   * The constructor instantiates an object of the code class with the current directory as one of
   * the parameters.
   *
   * @param currentDirectory the current directory
   */
  public Code(String currentDirectory) {
    if (currentDirectory == null || currentDirectory.isEmpty()) {
      throw new IllegalArgumentException("invalid current directory");
    }
    this.currentDirectory = currentDirectory;
    this.newDirectory = "";
    this.map = loadMap();
  }

  private Map<String, Runnable> loadMap() {
    Map<String, Runnable> temp = new HashMap<>();
    temp.put("", this::processEmpty);
    temp.put(".", this::processPeriod);
    temp.put("..", this::processParent);
    return temp;
  }

  /**
   * Processes the command to change the current directory to a new directory.
   *
   * @param cString the command to change the directory
   * @return the new directory
   */
  public String process(String cString) {
    if (cString.equals("")) {
      this.newDirectory = this.currentDirectory;
      return newDirectory;
    }
    cString = setNewDirectory(cString);
    String[] temp = new String[]{"", cString};
    String[] list = cString.contains("/") ? cString.split("/") : temp;
    for (int i = 1; i < list.length; i++) {
      if (map.containsKey(list[i])) {
        map.get(list[i]).run();
      } else {
        if (!(list[i].matches("[a-zA-Z0-9]*"))) {
          throw new IllegalArgumentException("No such file or Directory exists");
        }
        this.processDirectory(list[i]);
      }
    }
    return newDirectory;
  }

  private String setNewDirectory(String cString) {
    if (cString.startsWith("../") || cString.equals("..")) {
      String[] list = this.currentDirectory.split("/");
      String tempDirectory = '/' + list[list.length - 1];
      int index = currentDirectory.indexOf(tempDirectory);
      this.newDirectory = currentDirectory.substring(0, index);
      if (!(newDirectory.contains("/"))) {
        newDirectory = "/";
      }
      cString = cString.replaceFirst("..", "");
    } else if (cString.startsWith("./") || cString.equals(".")) {
      newDirectory = currentDirectory;
      cString = cString.replaceFirst(".", "");
    } else if (cString.startsWith("/")) {
      newDirectory = "/";
    } else {
      newDirectory = currentDirectory;
    }
    return cString;
  }


  private void processEmpty() {
    return;
  }

  private void processPeriod() {
    return;
  }

  private void processParent() {
    String[] list = this.newDirectory.split("/");
    String directory = "";
    if (list.length == 0) {
      directory = "/";
      
    } else {
      directory = "/" + list[list.length - 1];
    }
    int index = newDirectory.indexOf(directory);
    this.newDirectory = newDirectory.substring(0, index + 1);
    if (!(newDirectory.contains("/"))) {
      newDirectory = "/";
    }
  }

  private void processDirectory(String directory) {
    if (!(newDirectory.charAt(newDirectory.length() - 1) == '/')) {
      newDirectory = newDirectory + '/';
    }
    newDirectory = newDirectory + directory;
  }

  public static void main(String[] args) {
    String currentDirectory = "";
    String commandString = "";
    currentDirectory = args[0];
    commandString = args[1];
    Code code = new Code(currentDirectory);
    System.out.println(code.process(commandString));
  }
}
